### PR TITLE
Membership subscriptions

### PR DESCRIPTION
**Pull request by vivangkumar**
Adds membership subscriptions.

However, to make this possible, there were a few changes that had to be made

- Subscription resolution was arbitrary at first, however, with this addition we ensure that subscriptions are only returned when they resolve (either on first event or on open).
- Subscriptions that depend on each other are all waited on before the parent subscription is returned.
- Makes things slightly more consistent.
- Adds a `ResolvableSubscription` abstraction to make subscription resolution easier.
- Coroutines.

### CODE REVIEW

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/cursors/CursorService.kt [hunk 1]
```diff
@@ -50,44 +48,37 @@ internal class CursorService(private val chatManager: ChatManager) {
     }
 
     fun getReadCursor(userId: String, roomId: Int) : Future<Result<Cursor, Error>> =
-        (cursors[userId][roomId]?.asSuccess<Cursor, Error>() ?: notSubscribedToRoom("$roomId").asFailure())
+        (cursorsStore[userId][roomId]?.asSuccess<Cursor, Error>() ?: notSubscribedToRoom("$roomId").asFailure())
             .toFuture()
 
     private fun notSubscribedToRoom(name: String) =
-        Errors.other("Must be subscribed to room $name to access member's read cursors")
+        Errors.other("Must be subscribed to room $name to access member's read cursorsStore")
```
Comment (mdpye) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/cursors/CursorService.kt]: failure of Android Studio's ~context aware refactor~ naive search and replace tool

File: chatkit-core/src/main/kotlin/com/pusher/chatkit/CurrentUser.kt [hunk 2]
```diff
@@ -106,36 +107,42 @@ class CurrentUser(
...
+    ) = runBlocking {
+            chatManager.roomService.subscribeToRoom(roomId, consumer, messageLimit)
+                .autoRemove(roomId)
+                .also { roomSubscriptions += roomId to it }
+        }
+
+
+    private fun Subscription.autoRemove(roomId: Int) = object : ChatkitSubscription {
+        override suspend fun connect(): ChatkitSubscription {
+            return this
```
Comment (mdpye) [inline chatkit-core/src/main/kotlin/com/pusher/chatkit/CurrentUser.kt]: We should proxy this call to the object we're replacing, right?

**Comment by mdpye** [comment]
@vivangkumar I've rebased the prototyping commits we had out of the history because they were ultimately unused

**Comment by vivangkumar** [comment]
Sweet, thanks!

**Pull request closed without merging**